src/ledger/tests.zig

```diff
@@ -425,3 +487,130 @@ pub fn TestDB(scope: []const u8) type {
         }
     };
 }
+
+pub const BenchmarLegger = struct {
+    pub const min_iterations = 5;
+    pub const max_iterations = 5;
+
+    // Analogous to [bench_write_small](https://github.com/anza-xyz/agave/blob/cfd393654f84c36a3c49f15dbe25e16a0269008d/ledger/benches/blockstore.rs#L59)
+    pub fn benchWriteSmall() !u64 {
+        const allocator = std.heap.c_allocator;
+        var state = try State.initBench("bench write small");
+        defer state.deinitBench();
+        var inserter = try state.shredInserter();
+
+        const prefix = "agave.blockstore.bench_write_small.";
+        const shreds = try benchShreds(prefix ++ "shreds.bin");
+        defer inline for (.{shreds}) |slice| {
+            deinitShreds(allocator, slice);
+        };
+
+        const is_repairs = try inserter.allocator.alloc(bool, shreds.len);
+        defer inserter.allocator.free(is_repairs);
+        for (0..shreds.len) |i| {
+            is_repairs[i] = false;
+        }
+
+        var timer = try std.time.Timer.start();
+        _ = try inserter.insertShreds(shreds, is_repairs, null, false, null);
+        return timer.read();
+    }
+
+    // Analogous to [bench_read_sequential]https://github.com/anza-xyz/agave/blob/cfd393654f84c36a3c49f15dbe25e16a0269008d/ledger/benches/blockstore.rs#L78
+    pub fn benchReadSequential() !u64 {
+        const allocator = std.heap.c_allocator;
+        var state = try State.initBench("bentch read sequential");
+        defer state.deinitBench();
+        var inserter = try state.shredInserter();
+        var reader = try state.reader();
+
+        const prefix = "agave.blockstore.bench_read.";
+        const shreds = try benchShreds(prefix ++ "shreds.bin");
+        defer inline for (.{shreds}) |slice| {
+            deinitShreds(allocator, slice);
+        };
+
+        const total_shreds = shreds.len;
+
+        _ = try ledger.insert_shred.insertShredsForTest(&inserter, shreds);
+
+        const slot: u32 = 0;
+        const num_reads = total_shreds / 15;
+
+        var rng = std.Random.DefaultPrng.init(100);
+
+        var timer = try std.time.Timer.start();
+        const start_index = rng.random().intRangeAtMost(u32, 0, @intCast(total_shreds));
+        for (start_index..start_index + num_reads) |i| {
+            const shred_index = i % total_shreds;
+            _ = try reader.getDataShred(slot, shred_index);
+        }
+        return timer.read();
+    }
+
+    // Analogous to [bench_read_random]https://github.com/anza-xyz/agave/blob/92eca1192b055d896558a78759d4e79ab4721ff1/ledger/benches/blockstore.rs#L103
+    pub fn benchReadRandom() !u64 {
```

Comment:
```
Benchmark                  Iterations    Min(us)    Max(us)   Variance   Mean(us)
---------------------------------------------------------------------------------
benchReadRandom                     5    1993658    2326961 12944783078    2120888
```

Agave at 9c2098450ca7e5271e3690277992fbc910be27d0
```
running 1 test
test bench_read_random             ... bench:   2,820,841.70 ns/iter (+/- 28,719.65)
```

Reply:
With releasemall, I also got 

```
benchReadRandom                     5      53996      72074   38589915      60359
```

> it looks like benchReadSequential and benchReadRandom use the same input shreds in agave.blockstore.bench_read.shreds.bin. Is that intentional? Maybe each benchmark should use shreds that were generated by the respective agave benchmark, to ensure the data is comparable.

I'll look into this.

Reply:
This is because the setup code for bother are identical. (As far as I can see)

https://github.com/anza-xyz/agave/blob/9c2098450ca7e5271e3690277992fbc910be27d0/ledger/benches/blockstore.rs#L79-L88

and 

https://github.com/anza-xyz/agave/blob/9c2098450ca7e5271e3690277992fbc910be27d0/ledger/benches/blockstore.rs#L104-L113

Reply:
Do you know what's causing this discrepancy?

> I noticed that the agave benchmark reads 4369 shreds, whereas the sig benchmark reads 10499 indices. But that wouldn't explain a 50x slowdown. I also saw that in the agave benchmark, they only read 1/15 of the total number of shreds. Whereas in the sig benchmark, we read all of the shreds. But 4369 is not 1/15 of 10499, so something isn't adding up here.

Reply:
I believe the discrepancy should be corrected with this commit [2a1dbc6](https://github.com/Syndica/sig/pull/275/commits/2a1dbc66cbd0e647c718717447889895690c8017)

Running now, the result is:

```
Benchmark                  Iterations    Min(us)    Max(us)   Variance   Mean(us)
---------------------------------------------------------------------------------
benchReadRandom                     5       1180       1483      11685       1276

```

Reply:
With iteration as 1000

```
zig build -Doptimize=ReleaseSafe benchmark -- ledger
filtering benchmarks with prefix: ledger

Benchmark                  Iterations    Min(ns)    Max(ns)   Variance   Mean(ns)
---------------------------------------------------------------------------------
benchReadRandom                  1000     999917    4729250 59492207936    1295729
```